### PR TITLE
Support for `<col>` elements in the XML-Dom with according `align`

### PR DESCRIPTION
**Pull request by ThorbenSchiller**
# Description

This pull request includes functionality for <col> elements. The current implementation does not insert <col> elements by default for new tables, therefore customizing of those elements is not possible. To enable support, a new option `shouldCreateColumnSpecificationNodes` is introduced. If set to true, `<col>` elements will be created by default in the XML-DOM (if the schema supports this).

Furthermore, the attributes `align` and `valign` are allowed on `<col>` elements. Although this attributes are obsolete according to [HTML5](https://www.w3.org/TR/2012/WD-html-markup-20120329/col.html), align may be allowed according to XHTML table models. 

For visualization, I rely on output classes (for my use case only `align` is relevant)

configureSxModule.js:
```javascript
var availableAligns = ["center", "right", "left", "justify"];
availableAligns.forEach(align => {
    // add according class to cols which are aligned and have no align on itself
    const alignXPath = `let $current := self::td[not(@align)] return $current/ancestor::table[1]/col[position() = count($current/preceding-sibling::td) + 1 and @align="${align}"]`;
    configureProperties(sxModule, alignXPath, {
        outputClass: "table-cell-" + align
    });
});
```
The align only applies to standard cells (`<td>`), although it may be simple to change the xpath expression.

And the according css:
```css
[cv-output-class="table-cell-center"] {
    text-align: center;
}

[cv-output-class="table-cell-left"] {
    text-align: left;
}

[cv-output-class="table-cell-right"] {
    text-align: right;
}

[cv-output-class="table-cell-justify"] {
    text-align: justify;
}
```

The code follows the official API according to [configureAsTableElements](http://documentation.fontoxml.com/editor/latest/configure-a-custom-table-definition-12425222.html#id-..Configureacustomtabledefinitionv7.1.0v7.3.0-Configuringthetabledefinition).

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Unfortunately, the provided test pipeline is not accessible without granted access. The code has been tested in our customer implementation via manuell tests.

**Test Configuration**:
* SDK: 7.3.2

You may take a look at https://bitbucket.org/liones/fontoxml-app-werkml-e-spirit-editor for an example.

# Checklist:

- [ ] My code follows the style guidelines of this project (no coding conventions provided).
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] XPath Expression have been optimized for performance

### CODE REVIEW

**Comment by DrRataplan** [comment]
Hey Thorben,

Thanks a lot for this PR! :tada: we will review it as soon as possible!

I see that it's a bit impossible to actually run the unit tests if you're not a Fonto dev. This makes implementing these things way harder... @bwrrp, do you have some thoughts on this?

File: src/table-definition/XhtmlTableDefinition.js [hunk 1]
```diff
@@ -27,6 +29,7 @@ define([
 		var useTbody = !!options.useTbody;
 		var useTh = !!options.useTh;
 		var useBorders = options.useBorders !== false;
+		var shouldCreateColumnSpecificationNodes = options.shouldCreateColumnSpecificationNodes !== false;
```
Comment (DrRataplan) [inline src/table-definition/XhtmlTableDefinition.js]: I think you meant `!!options.shouldCreateColumnSpecificationNodes` to retain the current behaviour if nothing changed in the configuration on the app level

File: src/operations-column-alignment.json [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			{
+				"type": "transform/setContextNodeIdToSelectionAncestor",
+				"data": {
+					"selectionAncestorMatchesSomeOf": ["self::td", "self::th"]
+				}
+			},
+			{
+				"type": "transform/setContextNodeIdToFirstMatchingNodeFromContextNode",
+				"data": {
+					"xPathQuery": "let $current := . return $current/ancestor::table[1]/col[position() = count($current/preceding-sibling::td) + 1]"
```
Comment (DrRataplan) [inline src/operations-column-alignment.json]: This is a bit dangerous due to colspan / rowspan. Fonto has a [fonto:get-column-index](http://documentation.fontoxml.com/api/latest/fonto-get-column-index-12423499.html) XPath function which gives you the column in which the cell resides (0-based).

File: src/operations-column-alignment.json [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"_xhtml-set-column-align": {
+		"steps": [
+			{
+				"type": "transform/setContextNodeIdToSelectionAncestor",
+				"data": {
+					"selectionAncestorMatchesSomeOf": ["self::td", "self::th"]
+				}
+			},
+			{
+				"type": "transform/setContextNodeIdToFirstMatchingNodeFromContextNode",
```
Comment (DrRataplan) [inline src/operations-column-alignment.json]: Have you thought of using inline XPaths instead of the `setContextNodeIdToFirstMatchingNodeFromContextNode` transform? This is the `x__` prefix you might have seen in other operations. It is documented on the [inline XPath block of the docs on operations](http://documentation.fontoxml.com/editor/latest/operations-3098734.html). It's usually a lot less to write.

File: src/table-definition/XhtmlTableDefinition.js [hunk 4]
```diff
@@ -85,6 +88,8 @@ define([
 
 			supportsBorders: useBorders,
 
+			shouldCreateColumnSpecificationNodes: shouldCreateColumnSpecificationNodes,
```
Comment (DrRataplan) [inline src/table-definition/XhtmlTableDefinition.js:92]: Hmm, I think you've found a quirk in the table-flow generic code. `shouldCreateColumnSpecificationNodes` is unconditionally set to whether or not the `createColumnSpecificationNodeStrategy` is present on the options.

I'll file an internal bug report to address this issue.

File: src/table-definition/XhtmlTableDefinition.js [hunk 5]
```diff
@@ -139,14 +144,17 @@ define([
...
 			// Specification
 			getTableSpecificationStrategies: useBorders ? [
 					getSpecificationValueStrategies.createGetValueAsBooleanStrategy('borders', './@border = "1"')
 				] : [],
 
 			getColumnSpecificationStrategies: [
-					getSpecificationValueStrategies.createGetValueAsStringStrategy('columnWidth', '"1*"')
+					getSpecificationValueStrategies.createGetValueAsStringStrategy('columnWidth', '"1*"'),
+					getSpecificationValueStrategies.createGetValueAsStringStrategy('horizontalAlignment', './@align'),
+					getSpecificationValueStrategies.createGetValueAsStringStrategy('verticalAlignment', './@valign')
```
Comment (DrRataplan) [inline src/table-definition/XhtmlTableDefinition.js]: I do not think this is even needed. Some of the purely visual aspects of parts of a table are saved in our abstract table model (the thing we are declaring here) because of backwards compatibility reasons. For instance, we have a setCellData operation with which you can set a number of properties which are then synced back to the XML. I think the outputclass approach you chose in the PR is even preferable to using celldata because it is better in describing where which data should be saved in the XML.

By the way, the next Fonto release is scheduled to contain a new alignment property for just about every family. Using this property, the need for custom CSS + outputclasses is removed. We can then even think about configuring alignment for cells in this package.

File: src/operations-column-alignment.json (hunk 3 above)
Comment (wvbe) [reply src/operations-column-alignment.json]: @DrRataplan, could you help out? If you use the `x__` stuff, I give up at the following:

```
{
    "type": "operation/set-attributes",
    "data": {
        "contextNodeId": "let $current :=  x__$data('contextNode') return $current/ancestor::table[1]/col[position() = fonto:get-column-index($current) + 1]",
        "attributes": {
            "align": "{{align}}"
        },
        "disabledWhenUnchanged": true
    }
}
```

Reminder to self:
- Put `fonto:get-column-index` somewhere in there
- Mention `x__` in documentation of `setContextNodeIdToFirstMatchingNodeFromContextNode`
- Add code examples for `x__`
- Buy @DrRataplan a beer and think on how we can make the xpath playground more suitable to test this type of stuff

Comment (DrRataplan) [reply src/operations-column-alignment.json]: That XPath looks just about alright, but you seem to be missing the `x__` prefix. I'd suggest the following:
```
	"_xhtml-set-column-align": {
		"steps": [
			{
				"type": "transform/setContextNodeIdToSelectionAncestor",
				"data": {
					"selectionAncestorMatchesSomeOf": ["self::td", "self::th"]
				}
			},
			{
				"type": "operation/set-attributes",
				"data": {
					"contextNodeId": "x__let $current := $data('contextNode') return $current/ancestor::table[1]/col[position() = fonto:get-column-index($current) + 1]",
					"attributes": {
						"align": "{{align}}"
					},
					"disabledWhenUnchanged": true
				}
			}
		]
	}
```

Comment (DrRataplan) [reply src/operations-column-alignment.json]: Note to self: this code assumes the app binds the empty prefix to whatever namespace uri these tables reside in. IE. apps who wish to use the `html` prefix can not work with this code.

Because other operations in this package also make this assumption, it is not that bad though. We do not have to fix this at this moment.

Comment (ThorbenSchiller) [reply src/operations-column-alignment.json]: Hi @wvbe, @DrRataplan,

I tried similar in my implementation, but I can't get it to work with only `operation/set-attributes`. For me it seems like the operation does not have the correct `contextNode` (or any at all?). Once I set the context node in an additional step beforehand, it works for me:
```json
"steps": [
    {
        "type": "transform/setContextNodeIdToSelectionAncestor",
        "data": {
            "selectionAncestorMatchesSomeOf": ["self::td", "self::th"]
        }
    },
    {
        "type": "operation/set-attributes",
        "data": {
            "contextNodeId": "x__let $current := $data('contextNode') return $current/ancestor::table[1]/col[position() = fonto:get-column-index($current) + 1]",
            "attributes": {
                "align": "{{align}}"
            },
            "disabledWhenUnchanged": true
        }
    }
]
```

File: src/table-definition/XhtmlTableDefinition.js (hunk 1 above)
Comment (ThorbenSchiller) [reply src/table-definition/XhtmlTableDefinition.js]: Hi @DrRataplan,
oh yes, my bad. Thank you for pointing out :)

File: src/operations-column-alignment.json (hunk 3 above)
Comment (DrRataplan) [reply src/operations-column-alignment.json]: Yup, correct. That operation looks a lot more concise and should work like a charm.

File: src/table-definition/XhtmlTableDefinition.js (hunk 5 above)
Comment (ThorbenSchiller) [reply src/table-definition/XhtmlTableDefinition.js]: Hi @DrRataplan,
hm, sorry but I can't find the `setCellData` function in the [documentation](http://documentation.fontoxml.com/api/latest/function-12423104.html) (although it seems to be in the `fontoxml-table-flow` package). Is this an official API functionality?

Thanks for teasing me about the new Fonto update! Looking forward to it :)

Comment (DrRataplan) [reply src/table-definition/XhtmlTableDefinition.js]: Ah, yes, `setCellData` is an operation which is scheduled to be made public API in the next release. It provides an abstraction over cell data in some circumstances and is used for cell borders in some apps.

For these purely visual aspects, the abstraction turned out to be confusing. I like your approach of directly working with XML way better!

**Comment by DrRataplan** [comment]
As I see it, there's only a single point of feedback that needs to be resolved before integration: the columnspecification code. This can be removed because the platform only uses it for column width at this moment. Styling the individual cells based on the column works way better.

After integration, I'll make sure the unit tests work and the internal bug is resolved. That way, this code will land in 7.4 (if all goes well).

In 7.4, you can replace the outputclass + custom CSS with a new public API which directly sets the alignment of cells.

@thorb3N, I like your pull request format. Would you mind if I copy it as a PR format for the other Fonto repos?

**Comment by ThorbenSchiller** [comment]
@DrRataplan feel free. tbh I googled it and copied it from an [exisiting template](https://embeddedartistry.com/blog/2017/8/4/a-github-pull-request-template-for-your-projects). 
Also, it would be cool if you could add your coding guidelines and your code style (for auto format).

**Comment by DrRataplan** [comment]
Noted, thanks! And running the unit tests should be cool too, I'd imagine.

I've created a number of internal issues addressing these points.

Thanks for your feedback!

File: src/table-definition/XhtmlTableDefinition.js (hunk 4 above)
Comment (DrRataplan) [reply src/table-definition/XhtmlTableDefinition.js:92]: Note to self: this is DEV-1517 and has been committed in fontoxml-table-flow

**Comment by DrRataplan** [comment]
I have added a number of unit tests and I found a small edge-case with tables with only `<tr>` elements. I'll integrate this now.

Thanks again for this pull request!!!

**Review by DrRataplan — APPROVED** [review]
